src/ddspdrum/torchmodule.py

```diff
@@ -435,3 +439,60 @@ def __init__(
 
     def oscillator(self, argument):
         return torch.cos(argument)
+
+
+class TorchFmVCO(TorchVCO):
+    """
+    Frequency modulation VCO. Takes `mod_signal` as instantaneous frequency.
+
+    Typical modulation is calculated in pitch-space (midi). For FM to work,
+    we have to change the order of calculations. Here `mod_depth` is interpreted
+    as the "modulation index" which is tied to the fundamental of the oscillator
+    being modulated:
+
+        modulation_index = frequency_deviation / modulation_frequency
+
+    """
+
+    def __init__(
+            self,
+            midi_f0: float = 10.0,
+            mod_depth: float = 50.0,
+            phase: float = 0.0):
+        super().__init__(midi_f0=midi_f0, mod_depth=mod_depth, phase=phase)
+
+    def make_control_as_frequency(self, mod_signal: T):
+        # Compute modulation in Hz space (rather than midi-space).
+        f0_hz = midi_to_hz(self.p("pitch"))
+        fm_depth = self.p("mod_depth") * f0_hz
+        modulation_hz = fm_depth * mod_signal
+        return f0_hz + modulation_hz
+
+    def oscillator(self, argument):
+        # Classically, FM operators are sine waves.
+        return torch.cos(argument)
+
+
+# TODO: TorchSquareSawVCO
+
+
+class TorchVCA(TorchSynthModule):
+    """
+    Voltage controlled amplifier.
+    """
+
+    def __init__(
+            self,
+            sample_rate: int = SAMPLE_RATE,
+            buffer_size: int = BUFFER_SIZE
+    ):
+        super().__init__(sample_rate=sample_rate, buffer_size=buffer_size)
+
+    def _forward(self, control_in: T, audio_in: T) -> T:
+        assert (control_in >= 0).all() and (control_in <= 1).all()
+
+        if (audio_in <= -1).any() or (audio_in >= 1).any():
+            normalize(audio_in)
+
+        audio_in = fix_length(audio_in, len(control_in))
+        return control_in * audio_in
```

Comment:
This might be controversial, but needed to be addressed in both the npy and torch modules:

Before the VCA clipped both control and audio signals. Here, we assert the control is in the right range (unlike pitch mod, no reason for it to ever be negative; and no reason to be above 1). The audio is generally well-behaved, but sometimes can creep outside [-1, 1]. With most random synths this is not the case, but it's here as a fail safe. It shouldn't effect synth performance.